src/app/modules/crypto/reducers.ts

```diff
@@ -7,6 +7,7 @@ export interface CryptoState {
   password: null | string;
   testCipher: null | string;
   isRequestingPassword: boolean;
+  isChangingPassword: boolean;
```

Comment:
You are correct, it would be ideal to have this field in the component state. I struggled with the decision to put this in redux state, but this was the only way I could get the UX that I was aiming for. The user flow is:
1. Click Change Password button
2. Prompt for current password with the `PasswordPrompt` modal (drawer remains closed)
3. After password entered, open drawer displaying the new password form
4. Enter the new password twice then click Continue button
5. Set the new password in redux & browser.storage
6. Automatically close the drawer

If we use component state, then how would the `NodeSettings` know to open the drawer at step 3? Wouldn't it have to open the drawer immediately when Change Password is clicked? Then we would also have to figure out how to inform the `NodeSettings` if the user closes the `PasswordPrompt` without entering in their current password, so that the drawer can automatically closed. Also for step 6, how can we close the drawer automatically? First thought would be to use the same `if` condition used in `componentWillUpdate` which we both agree is poo. 

I'll work on moving `isChangingPassword` to component state and try to keep the code as clean as possible. If you have any suggestions on the implementation, I am open ears.

Reply:
Password prompt and salt generation could all be done _after_ the user enters their new password (CreatePassword onCreatePassword's prop action, instead of on button click.) This would also remove the need to set a new salt, and then revert if they cancel, as seen in the changePassword saga.

Reply:
You could auto-close drawer using a heuristic prop change (e.g. if prevProps.salt !== this.props.salt)

Reply:
Ah ok, I see what you are saying. But you don't think it's a bit weird to first ask a user to type in the new password twice, then to get prompted to enter their old password? The node url and macaroon forms work this way but I felt it necessary to change the flow for the password to prompt for the old one first. If you prefer to prompt after then I'll code it up this way, no prob.

Reply:
Yup, I think this 3 field approach is definitely the best way to do this. I just have one final question about what you think is the best way to do this. I can immediately think of 3 ways to achieve what we want considering the current `CreatePassword` component needs to behave the same in the onboarding process.

1. Add a boolean `requestCurrentPassword` prop to `CreatePassword` and when set to true, it will display the current password field on top. When the submit button is clicked, the component will call `this.props.onCreatePassword` with both the current and new passwords.
2. Add the current password form field in the `NodeSettings` component and render it above the `<CreatePassword />` component in the drawer. 
3. Create a new `ChangePassword` component that would basically be a copy of `CreatePassword` but it will have the 3 fields instead of two. I could probably refactor the password strength indicator into a shared component.

In all three cases, I'd need to add the logic to test the current password against the `TEST_CIPHER_DATA`.

Please let me know what you think is best.

Reply:
Both 1 & 2 sound like good approaches, I'll defer to you on what you think is best between the two.

Reply:
I used the first approach